test/run-drun/timer.mo

```diff
@@ -0,0 +1,201 @@
+            }
+    };
+
+    // ad-hoc place for the Timer.mo API
+    type TimerId = Nat;
+    func setTimer(delaySecs : Nat64, recurring : Bool, job : () -> async ()) : TimerId {
+        lastId += 1;
+        let id = lastId;
+        let now = time();
+        let delayNanos = 1_000_000_000 * delaySecs;
+        let expire = now + delayNanos;
+        let delay = if recurring ?delayNanos else null;
+        // only works on pruned nodes
+        func insert(n : ?Node) : Node =
+            switch n {
+              case null ({ var expire; id; delay; job; ante = null; dopo = null });
+              case (?n) {
+                 assert n.expire != 0;
+                 if (expire < n.expire) ({ n with ante = ?(insert(n.ante)) })
+                 else ({ n with dopo = ?(insert(n.dopo)) })
+               }
+            };
+        timers := ?insert(prune(timers));
+
+        let exp = nextExpiration timers;
+        if (exp == 0) timers := null;
+        let prev = (prim "global_timer_set" : Nat64 -> Nat64) exp;
+        /*FIXME: this is expensive*/
+        if (prev != 0 and prev != 0 and exp > prev) {
+            // reinstall
+            ignore (prim "global_timer_set" : Nat64 -> Nat64) prev;
+        };
+
+        //debug { debugPrint(debug_show {now; exp; prev; id }) };
+
+        id
+    };
+
+    func graft(onto : ?Node, branch : ?Node) : ?Node = switch (onto, branch) {
+        case (null, null) null;
+        case (null, _) branch;
+        case (_, null) onto;
+        case (?onto, _) { ?{ onto with dopo = graft(onto.dopo, branch) } }
+    };
+
+    func cancelTimer(id : TimerId) {
+        func hunt(n : ?Node) : ?Node = switch n {
+          case null n;
+          case (?{ id = node; ante; dopo }) {
+                   if (node == id) {
+                       graft(ante, dopo)
+                   } else do? {
+                       { n! with ante = hunt ante; dopo = hunt dopo }
+                   }
+               }
+        };
+
+        timers := hunt timers;
+
+        if (nextExpiration timers == 0) {
+            // no more expirations ahead
+            ignore (prim "global_timer_set" : Nat64 -> Nat64) 0;
+            timers := null
+        }
+    };
+
+
+  let raw_rand = (actor "aaaaa-aa" : actor { raw_rand : () -> async Blob }).raw_rand;
+
+  var count = 0;
+  var max = 10;
+
+  system func timer() : async () {
+    let now = time();
+
+    type CNode = ?{var expire : Nat64; id : TimerId; delay : ?Nat64; ante : CNode; dopo : CNode };
+    func clean(n : ?Node) : CNode =
+      switch n {
+          case null null;
+          case (?n) ?{n with ante = clean(n.ante); dopo = clean(n.dopo) }
+      };
+    //debugPrint(debug_show {now; timers = clean timers});
+
+    let exp = nextExpiration timers;
+    let prev = (prim "global_timer_set" : Nat64 -> Nat64) exp;
+
+    if (exp == 0) {
+        return
+    };
+
+    var gathered = 0;
+    let thunks : [var ?(() -> async ())] = Array_init(10/*FIXME*/, null);
```

Comment:
Either expand the array when filled or stop putting elements in it when full (and leave the expiration of surplus timers to the next firing of the global timer).